### PR TITLE
Windowing again

### PR DESCRIPTION
**Pull request by Licenser**
# Pull request

## Description

- clean up windowing code (better eviction)
- fix a bunch of bugs in the unit testing framework

## Related

* Related [docs PR](https://github.com/tremor-rs/tremor-www-docs/pull/176)

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

## Performance

Within run to run variance of main

### CODE REVIEW

File: tremor-script/src/ast.rs [hunk 1]
```diff
@@ -994,6 +1002,22 @@ pub enum Expr<'script> {
     Imut(ImutExprInt<'script>),
 }
 
+impl<'script> Expr<'script> {
+    /// Tries to borrow the Expor as an `Invoke`
```
Comment (mfelsche) [inline tremor-script/src/ast.rs:1006]:
```suggestion
    /// Tries to borrow the Expr as an `Invoke`
```

File: tremor-pipeline/src/op/trickle/select.rs [hunk 2]
```diff
@@ -603,2059 +214,261 @@ impl Operator for TrickleSelect {
...
 
-                                // get current window group
-                                let this_group = stry!(get_or_create_group(
-                                    window,
-                                    event_id_gen,
-                                    aggregates,
-                                    group_str,
-                                    group_value,
-                                ));
+            let ctx = EventContext::new(ingest_ns, None);
```
Comment (mfelsche) [inline tremor-pipeline/src/op/trickle/select.rs:382]: we should set the cartdinality here, right?

File: tremor-pipeline/src/op/trickle/select.rs [hunk 3]
```diff
@@ -603,2059 +214,261 @@ impl Operator for TrickleSelect {
...
-                                    std::mem::swap(&mut scratch1.aggregates, &mut this_group.aggrs);
-                                    scratch1.transactional = this_group.transactional;
-                                    // aggrs.init() and reset transactional state
-                                    this_group.reset();
-                                } else {
-                                    // add this event to the aggr state **AFTER** emit and propagate to next windows
+                    if window_event.emit {
+                        // push
+                        let mut env = env(&ctx, run, node_meta, recursion_limit);
+                        env.aggrs = dbg!(&w.aggrs);
```
Comment (mfelsche) [inline tremor-pipeline/src/op/trickle/select.rs:398]: 😱 leftover `dbg!`

File: tremor-pipeline/src/op/trickle/select.rs [hunk 4]
```diff
@@ -603,2059 +214,261 @@ impl Operator for TrickleSelect {
...
-                                    }
-                                    this_group.track_transactional(scratch1.transactional);
-                                    this_group.id.track(&event_id_scratch1);
-                                }
-                            }
-                        } else {
-                            // if this happens, we fucked up internally.
-                            return Err("Internal Error. Cannot execute select with multiple windows, no scratches available. This is bad!".into());
+                        if let Some(next) = &mut w.next {
+                            can_remove = next.on_event(
```
Comment (mfelsche) [inline tremor-pipeline/src/op/trickle/select.rs:429]: why calling `on_event` for `next` when it has not been called on `w`? Shouldnt we handle the signal case differently?

File: tremor-script/src/ast/query.rs [hunk 5]
```diff
@@ -107,7 +109,7 @@ pub struct SelectStmt<'script> {
     pub aggregates: Aggregates<'static>,
     /// scratches needed when executing multiple windows
     /// only necessary if we have multiple windows, otherwise empty
-    pub aggregate_scratches: Option<(AggregateScratch, AggregateScratch)>,
+    pub aggregate_scratches: (AggregateScratch, AggregateScratch),
```
Comment (mfelsche) [inline tremor-script/src/ast/query.rs:112]: i think those can be removed, as they are not used anywhere! 🥳

File: tremor-script/src/ast/query/raw.rs [hunk 6]
```diff
@@ -157,12 +153,12 @@ impl<'script> Upable<'script> for StmtRaw<'script> {
                     .collect();
                 // only allocate scratches if they are really needed - when we have multiple windows
                 let aggregate_scratches = if stmt.windows.len() > 1 {
```
Comment (mfelsche) [inline tremor-script/src/ast/query/raw.rs:155]: Kill it with :fire:

**Review by mfelsche — COMMENTED** [review]
Wow, that is a ton of stuff in here! The resulting select code looks so much better than what was there before. 💯 

Left some comments and questions :)

File: tremor-pipeline/src/op/trickle/select.rs (hunk 4 above)
Comment (Licenser) [reply tremor-pipeline/src/op/trickle/select.rs:429]: `w.on_event` would try to accumulate the data, which we don't have, on a select, that would mark the window as tracking data, adding empty values, emitting it, and keeping it around. for signal handling we needed a different logic for `on_event` that circumvent the ingest an event part.

File: tremor-pipeline/src/op/trickle/select.rs (hunk 2 above)
Comment (Licenser) [reply tremor-pipeline/src/op/trickle/select.rs:382]: Yes! Good catch :)

File: tremor-pipeline/src/op/trickle/select.rs (hunk 4 above)
Comment (mfelsche) [reply tremor-pipeline/src/op/trickle/select.rs:430]: Just wanted to check that this is on purpose and not a blind-spot. Happens with huge workloads like this PR (at least to me).

Comment (Licenser) [reply tremor-pipeline/src/op/trickle/select.rs:430]: It easily does :D yes, and sorry for the size :)

**Review by mfelsche — APPROVED** [review]
BAM!

**Pull request merged**